### PR TITLE
facts - Set correct OS family for Rocky Linux

### PR DESCRIPTION
**Pull request by rkm**
##### SUMMARY

This change adds support for [RockyLinux](https://rockylinux.org/) and adds it as a member to the RedHat family of operating systems.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

lib/ansible/module_utils/facts/system/distribution.py

##### ADDITIONAL INFORMATION

- Don't think we have a codename for this release, but looking at some other distros it seems like this isn't required

### CODE REVIEW

File: changelogs/fragments/support_rockylinux.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - add RockyLinux to fact gathering (https://github.com/ansible/ansible/pull/74530)
```
Comment (Akasurde) [inline changelogs/fragments/support_rockylinux.yml]:
```suggestion
  - Add RockyLinux to fact gathering (https://github.com/ansible/ansible/pull/74530).
```

**Review by Akasurde — CHANGES_REQUESTED** [review]

File: changelogs/fragments/support_rockylinux.yml (hunk 1 above)
Comment (Akasurde) [reply changelogs/fragments/support_rockylinux.yml]: Rest LGTM

**Review by Akasurde — APPROVED** [review]
LGTM, will wait for others to comment.

File: changelogs/fragments/support_rockylinux.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - Add RockyLinux to fact gathering (https://github.com/ansible/ansible/pull/74530).
```
Comment (Akasurde) [inline changelogs/fragments/support_rockylinux.yml:2]:
```suggestion
  - Add RockyLinux to fact gathering (https://github.com/ansible/ansible/pull/74530).
  - hostname - Add support for RockyLinux (https://github.com/ansible/ansible/pull/74530).
```

File: test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            "distributor_id": "RockyLinux",
+            "description": "RockyLinux release 8.3",
+            "release": "8.3",
+            "codename": ""
+        },
+        "os_release_info": {
+            "name": "RockyLinux",
+            "version": "8.3",
+            "id": "rockylinux",
+            "id_like": "rhel fedora",
```
Comment (imjoseangel) [inline test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json:14]:
```suggestion
            "id_like": "rhel centos fedora",
```

**Review by imjoseangel — COMMENTED** [review]
LGTM, I'm not sure if it could be good adding centos under id_like. What do you think?

**Comment by rkm** [comment]
> LGTM, I'm not sure if it could be good adding centos under id_like. What do you think?

I did ask about this, and was advised not to include it. Reason being that Rocky is derived from RHEL and not CentOS, so we should have the same `ID_LIKE` as CentOS.

File: lib/ansible/module_utils/facts/system/distribution.py [hunk 4]
```diff
@@ -480,7 +480,7 @@ class Distribution(object):
     OS_FAMILY_MAP = {'RedHat': ['RedHat', 'Fedora', 'CentOS', 'Scientific', 'SLC',
                                 'Ascendos', 'CloudLinux', 'PSBM', 'OracleLinux', 'OVS',
                                 'OEL', 'Amazon', 'Virtuozzo', 'XenServer', 'Alibaba',
-                                'EulerOS', 'openEuler', 'AlmaLinux'],
+                                'EulerOS', 'openEuler', 'AlmaLinux', 'RockyLinux'],
```
Comment (nazunalika) [inline lib/ansible/module_utils/facts/system/distribution.py]: Please change to 'Rocky' -- The ID is simply `Rocky`
```suggestion
                                'EulerOS', 'openEuler', 'AlmaLinux', 'Rocky'],
```

File: lib/ansible/modules/hostname.py [hunk 5]
```diff
@@ -1001,6 +1001,12 @@ class PopHostname(Hostname):
     strategy_class = DebianStrategy
 
 
+class RockyLinuxHostname(Hostname):
+    platform = 'Linux'
+    distribution = 'RockyLinux'
```
Comment (nazunalika) [inline lib/ansible/modules/hostname.py]: Please change to 'Rocky' -- The ID is simply `Rocky`
```suggestion
    distribution = 'Rocky'
```

File: test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json (hunk 3 above)
Comment (nazunalika) [inline test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json]: Please match the json fixtures according to `/etc/os-release`. Rocky Linux is separated. id_like is `rhel fedora`
```suggestion
    "name": "Rocky Linux 8.3",
    "distro": {
        "codename": "",
        "id": "rocky",
        "name": "Rocky Linux",
        "version": "8.3",
        "version_best": "8.3",
        "lsb_release_info": {
            "lsb_version": ":core-4.1-amd64:core-4.1-noarch",
            "distributor_id": "Rocky",
            "description": "Rocky Linux release 8.3",
            "release": "8.3",
            "codename": ""
        },
        "os_release_info": {
            "name": "Rocky Linux",
            "version": "8.3",
            "id": "rocky",
            "id_like": "rhel fedora",
```

```
NAME="Rocky Linux"
VERSION="8"
ID="rocky"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Rocky Linux 8"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:rocky:rocky:8"
HOME_URL="https://rockylinux.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
ROCKY_SUPPORT_PRODUCT="Rocky Linux"
ROCKY_SUPPORT_PRODUCT_VERSION="8"
```

File: test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    ],
+    "result": {
+        "distribution": "RockyLinux",
+        "distribution_version": "8.3",
+        "distribution_release": "NA",
+        "distribution_major_version": "8",
+        "os_family": "RedHat"
+    },
+    "platform.release": "4.18.0-240.22.1.el8.x86_64"
+}
```
Comment (nazunalika) [inline test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json]: Please match `/etc/os-release` and `/etc/redhat-release`
```suggestion
            "version_id": "8.3",
            "platform_id": "platform:el8",
            "pretty_name": "Rocky Linux 8.3",
            "ansi_color": "0;34",
            "cpe_name": "cpe:/o:rockylinux:rockylinux:8.3",
            "home_url": "https://rockylinux.org/",
            "bug_report_url": "https://bugs.rockylinux.org/",
            "rocky_support_product": "Rocky Linux",
            "rocky_support_product_version": "8",
            "codename": ""
        }
    },
    "input": {
        "/etc/rocky-release": "Rocky Linux release 8.3\n",
        "/etc/redhat-release": "Rocky Linux release 8.3\n",
        "/etc/system-release": "Rocky Linux release 8.3\n",
        "/etc/os-release": "NAME=\"Rocky Linux\"\nVERSION=\"8\"\nID=\"rocky\"\nID_LIKE=\"rhel fedora\"\nVERSION_ID=\"8\"\nPLATFORM_ID=\"platform:el8\"\nPRETTY_NAME=\"Rocky Linux 8\"\nANSI_COLOR=\"0;31\"\nCPE_NAME=\"cpe:/o:rocky:rocky:8\"\nHOME_URL=\"https://rockylinux.org/\"\nBUG_REPORT_URL=\"https://bugs.rockylinux.org/\"\nROCKY_SUPPORT_PRODUCT=\"Rocky Linux\"\nROCKY_SUPPORT_PRODUCT_VERSION=\"8\"\n",
        "/usr/lib/os-release": "NAME=\"Rocky Linux\"\nVERSION=\"8\"\nID=\"rocky\"\nID_LIKE=\"rhel fedora\"\nVERSION_ID=\"8\"\nPLATFORM_ID=\"platform:el8\"\nPRETTY_NAME=\"Rocky Linux 8\"\nANSI_COLOR=\"0;31\"\nCPE_NAME=\"cpe:/o:rocky:rocky:8\"\nHOME_URL=\"https://rockylinux.org/\"\nBUG_REPORT_URL=\"https://bugs.rockylinux.org/\"\nROCKY_SUPPORT_PRODUCT=\"Rocky Linux\"\nROCKY_SUPPORT_PRODUCT_VERSION=\"8\"\n"
    },
    "platform.dist": [
        "rockylinux",
        "8.3",
        ""
    ],
    "result": {
        "distribution": "Rocky Linux",
        "distribution_version": "8.3",
        "distribution_release": "NA",
        "distribution_major_version": "8",
        "os_family": "RedHat"
    },
    "platform.release": "4.18.0-240.22.1.el8.x86_64"
}
```

```
NAME="Rocky Linux"
VERSION="8"
ID="rocky"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Rocky Linux 8"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:rocky:rocky:8"
HOME_URL="https://rockylinux.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
ROCKY_SUPPORT_PRODUCT="Rocky Linux"
ROCKY_SUPPORT_PRODUCT_VERSION="8"
```

```
Rocky Linux release 8.3
```

**Review by nazunalika — CHANGES_REQUESTED** [review]
Hello. I am from Release Engineering for Rocky Linux. I have reviewed the changes in this PR and they are incorrect and do not match our `/etc/os-release` nor `/etc/redhat-release` files. I agree that `ID_LIKE` should stay `"rhel fedora"`. 

With that said, the OS should be named "Rocky". According to our `/etc/os-release` and `/etc/rocky-release`:

```
# /etc/os-release
NAME="Rocky Linux"
VERSION="8"
ID="rocky"
ID_LIKE="rhel fedora"
VERSION_ID="8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Rocky Linux 8"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:rocky:rocky:8"
HOME_URL="https://rockylinux.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
ROCKY_SUPPORT_PRODUCT="Rocky Linux"
ROCKY_SUPPORT_PRODUCT_VERSION="8"
```

```
# /etc/rocky-release
Rocky Linux release 8.3
```

Please ensure the changes have been addressed.

**Comment by Akasurde** [comment]
@rkm Could you please generate output using https://github.com/ansible/ansible/blob/devel/hacking/tests/gen_distribution_version_testcase.py? and update here.

**Review by Akasurde — CHANGES_REQUESTED** [review]
As per above.

**Comment by rkm** [comment]
Done! Thanks for reviewing @nazunalika

File: lib/ansible/modules/hostname.py [hunk 7]
```diff
@@ -1001,6 +1001,12 @@ class PopHostname(Hostname):
     strategy_class = DebianStrategy
 
 
+class RockyLinuxHostname(Hostname):
```
Comment (nazunalika) [inline lib/ansible/modules/hostname.py]:
```suggestion
class RockyHostname(Hostname):
```

As I noted in the Rocky mattermost, we should probably name it like this for consistency.

**Review by nazunalika — CHANGES_REQUESTED** [review]

**Comment by Akasurde** [comment]
@nazunalika @rkm test failing at -

```
02:28 E           AssertionError: Comparing value of distribution on Rocky Linux 8.3, should: Rocky Linux, is: Rocky
02:28 E           assert 'Rocky' == 'Rocky Linux'
02:28 E             - Rocky
02:28 E             + Rocky Linux
02:28 

```

File: test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json [hunk 8]
```diff
@@ -0,0 +1,53 @@
+{
+    "name": "Rocky Linux 8.3",
```
Comment (nazunalika) [inline test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json]:
```suggestion
    "name": "Rocky 8.3",
```

**Comment by rkm** [comment]
@nazunalika I don't have time until later today to fixup those tests. I've added you as a contributor to my fork if you'd like to take a look

File: test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json [hunk 9]
```diff
@@ -0,0 +1,46 @@
+{
+    "name": "Rocky 8.3",
+    "distro": {
+        "codename": "",
+        "id": "rocky",
+        "name": "Rocky Linux",
+        "version": "8.3",
```
Comment (samdoran) [inline test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json:7]: In my testing `distro.version()` returns `8` not `8.3`. This means `result['distribution_version']` will be `8` and not `8.3`.

```
{
    "name": "Rocky 8",
    "distro": {
        "codename": "",
        "id": "rocky",
        "name": "Rocky Linux",
        "version": "8",
        "version_best": "8.3",
        "lsb_release_info": {},
        "os_release_info": {
            "name": "Rocky Linux",
            "version": "8",
            "id": "rocky",
            "id_like": "rhel fedora",
            "version_id": "8",
            "platform_id": "platform:el8",
            "pretty_name": "Rocky Linux 8",
            "ansi_color": "0;31",
            "cpe_name": "cpe:/o:rocky:rocky:8",
            "home_url": "https://rockylinux.org/",
            "bug_report_url": "https://bugs.rockylinux.org/",
            "rocky_support_product": "Rocky Linux",
            "rocky_support_product_version": "8"
        }
    },
    "input": {
        "/etc/redhat-release": "Rocky Linux release 8.3\n",
        "/etc/system-release": "Rocky Linux release 8.3\n",
        "/etc/os-release": "NAME=\"Rocky Linux\"\nVERSION=\"8\"\nID=\"rocky\"\nID_LIKE=\"rhel fedora\"\nVERSION_ID=\"8\"\nPLATFORM_ID=\"platform:el8\"\nPRETTY_NAME=\"Rocky Linux 8\"\nANSI_COLOR=\"0;31\"\nCPE_NAME=\"cpe:/o:rocky:rocky:8\"\nHOME_URL=\"https://rockylinux.org/\"\nBUG_REPORT_URL=\"https://bugs.rockylinux.org/\"\nROCKY_SUPPORT_PRODUCT=\"Rocky Linux\"\nROCKY_SUPPORT_PRODUCT_VERSION=\"8\"\n",
        "/usr/lib/os-release": "NAME=\"Rocky Linux\"\nVERSION=\"8\"\nID=\"rocky\"\nID_LIKE=\"rhel fedora\"\nVERSION_ID=\"8\"\nPLATFORM_ID=\"platform:el8\"\nPRETTY_NAME=\"Rocky Linux 8\"\nANSI_COLOR=\"0;31\"\nCPE_NAME=\"cpe:/o:rocky:rocky:8\"\nHOME_URL=\"https://rockylinux.org/\"\nBUG_REPORT_URL=\"https://bugs.rockylinux.org/\"\nROCKY_SUPPORT_PRODUCT=\"Rocky Linux\"\nROCKY_SUPPORT_PRODUCT_VERSION=\"8\"\n"
    },
    "platform.dist": [
        "rocky",
        "8",
        ""
    ],
    "result": {
        "distribution": "Rocky",
        "distribution_version": "8",
        "distribution_release": "NA",
        "distribution_major_version": "8",
        "os_family": "Rocky"
    },
    "platform.release": "4.18.0-240.22.1.el8.x86_64"
}
```

`rocky` needs to be added to the list of distros that need the more precise version information, and the fixture needs updating.
https://github.com/ansible/ansible/blob/d8c20a73a4c3d7d0383167a2c50dd077d15b1ba8/lib/ansible/module_utils/common/sys_info.py#L53-L56

This conditional needs to be updated as well.
https://github.com/ansible/ansible/blob/d8c20a73a4c3d7d0383167a2c50dd077d15b1ba8/lib/ansible/module_utils/common/sys_info.py#L63-L76

**Review by samdoran — CHANGES_REQUESTED** [review]
The facts changes need some further work. The `hostname` changes work correctly but should be in a separate PR.

File: test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json (hunk 9 above)
Comment (nazunalika) [reply test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json:7]: This does not need further changes if you update your release to the latest version. The "precise" versions are addressed by adding the minor release to `/etc/os-release` which RHEL does, but centos does not do.

Please update to the [latest version](https://kojidev.rockylinux.org/koji/buildinfo?buildID=7923) if you wish and test. We have not pushed updates to our repos yet.

Thank you.

**Comment by nazunalika** [comment]
I have opened #74545 for the hostname and updated this PR just for adding Rocky Linux to the RedHat class of distributions.

File: test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json (hunk 9 above)
Comment (samdoran) [reply test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json:7]: Awesome! Thanks. I just tested with a recent RC, but evidently it was not the latest.

Comment (samdoran) [reply test/units/module_utils/facts/system/distribution/fixtures/rockylinux_8_3.json:7]: I tested with `rocky-release-8.3-13.el8.noarch` and that gives the expected output.

**Review by samdoran — APPROVED** [review]

**Pull request merged**